### ISSUE TITLE
Roslyn fails to learn about variables' possible nullability from pure null checks in pattern matching over tuple expressions

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#51853]
**Version Used**: 
```
Branch main (8 Mar 2021)
Latest commit eddde7c by Sam Harwell:
Merge pull request #51559 from sharwell/current-thread

Use Environment.CurrentManagedThreadId
```

**Steps to Reproduce**:

Compile the following code
```cs
#nullable enable
public class C
{
    int Test(object y) => 1;

    void M1(object x, object y)
    {
        _ = (x, y) switch
        {
            (not null, null) => 1,
            (not null, { }) => 2,
            _ => Test(y), // no warnings
        };
    }
    
    void M2(object x, object y)
    {
        var t = (x, y);
        _ = t switch
        {
            (not null, null) => 1,
            (not null, { }) => 2,
            _ => Test(t.y), // CS8604: Possible null reference argument
        };
    }
}
```

**Expected Behavior**:
`CS8604` is reported in both methods since they explicitly test the respective argument for `null` before using it

**Actual Behavior**:
`CS8604: Possible null reference argument` is reported for the tuple variable in `M2` but not for the tuple expression in `M1`

**Notes**
I tried to debug this scenario and it seems that there are two separate problems that prevent this from working
- First, since `VisitSwitchExpressionCore` requests the governing expression slot before vising it, the placeholder slot for a tuple expression is not yet allocated when `VisitSwitchExpressionCore` requests it to learn from null patterns
- Second, if the slot _is_ allocated due to multiple analysis passes (e.g. if you add `for(;;)` before the switch expression) the variable used to learn from null patterns is a placeholder's field slot `(x, y).Item2` which doesn't propagate this knowledge to `y` and is effectively lost since it's impossible to reuse this field
![image](https://user-images.githubusercontent.com/5566206/111009803-ac63db80-83a5-11eb-8b1c-422e3a376049.png)
